json_schema_fuzz/schema_operations.py

```diff
@@ -1,45 +1,245 @@
 """ Operations on schemas """
 import copy
+import itertools
+import math
 from typing import Any, Dict, List
 
 
+def lcm(numbers):
+    """
+    Find least common multiple of a list of numbers
+
+    TODO replace with math.lcm after updating to Python 3.9
+    """
+    product = 1
+    gcd = 1
+    for num in numbers:
+        gcd = math.gcd(gcd, num)
+        product *= num
+    return product // gcd
+
+
+def get_val_or_none(dictionaries, key):
+    """
+    Get list of values from dictionaries.
+    If it is not present in any dictionary, return None.
+    """
+    values = []
+    for dictionary in dictionaries:
+        value = dictionary.get(key, None)
+        if value is not None:
+            values.append(value)
+    if len(values) > 0:
+        return values
+    else:
+        return None
+
+
+def get_index_or_default(given_list, index, default=None):
+    """ Get index from a list or get a default value """
+    try:
+        return given_list[index]
+    except IndexError:
+        return default
+
+
+def all_equal(given_list):
+    """ Check every array element is equal """
+    return given_list.count(given_list[0]) == len(given_list)
+
+
+class SchemaIsFalse(Exception):
+    """ Raised if schema will not validate true for any values """
+
+
+# pylint: disable=too-many-branches
+# pylint: disable=too-many-locals
+# pylint: disable=too-many-statements
 def merge(
-    schema_a: Dict[Any, Any],
-    schema_b: Dict[Any, Any],
-    path: List[Any] = None,
-    in_place: bool = False,
-):
-    """Merge two JSON schemas recursively."""
-    if path is None:
-        path = []
-
-    if not in_place:
-        # Make a copy of a so that it doesn't get modified in place
-        schema_a = copy.deepcopy(schema_a)
-
-    for key in schema_b:
-        if key not in schema_a:
-            schema_a[key] = schema_b[key]
-            continue
-        if isinstance(schema_a[key], dict) and isinstance(schema_b[key], dict):
-            merge(schema_a[key], schema_b[key],
-                  path=path + [str(key)],
-                  in_place=True)
-        elif schema_a[key] == schema_b[key]:
-            pass  # same leaf value
-        elif (
-                isinstance(schema_a[key], list)
-                and isinstance(schema_b[key], list)
-        ):
-            # Append lists together
-            schema_a[key].extend(schema_b[key])
+    *schemas: List[Dict[Any, Any]],
+) -> Dict[Any, Any]:
+    """
+    Merge a list of JSON schemas recursively.
+
+    Merging schemas here means that all schemas
+    must be true for result schema to be true.
```

Comment:
What does it mean for a schema to "be true"?

`merge(a, b)` is  schema such that an instance validates against `merge(a, b)` if and only if it validates against both `a` and `b`. In other words, `merge(a, b)` is functionally equivalent to `{"allOf": [a, b]}`, but without using `allOf`.